### PR TITLE
Implementation of Lemire's nearly divisionless method

### PR DESCRIPTION
**Pull request by mla-alm**
https://github.com/dotnet/runtime/issues/75395

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=79790) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=79790) before we can accept your contribution.<br/><br/>:x: mla-alm [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=79790)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=79790) it.</sub>

**Comment by danmoseley** [comment]
Could you please add benchmark(s) to dotnet/performance so we protect your work when it goes in? Or is coverage good enough already?

**Comment by mla-alm** [comment]
Which numbers or intervals to pick for benchmarks decides which algorithm is "optimal". IMO it could be interesting to include the edge cases:
Next(int.MinValue, 1)
NextInt64(long.MinValue, 1)
However these would actually point towards an implementation like https://github.com/apple/swift/pull/39143/commits/87b3f607042e653a42b505442cc803ec20319c1c.
For this implementation I guess the benchmarks are sufficient.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs [hunk 1]
```diff
@@ -28,6 +28,11 @@ internal sealed class XoshiroImpl : ImplBase
             //
             //     See <http://creativecommons.org/publicdomain/zero/1.0/>.
 
+            // Next(int maxValue), Next(int minValue, int maxValue), NextInt64(long maxValue) and NextInt64(int minValue, int maxValue)
+            // are based on the algorithm from https://arxiv.org/pdf/1805.10941.pdf:
+            //
+            //     Written in 2018 by Daniel Lemire
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:34]: could you please add a section in the THIRD-PARTY-NOTICES.TXT at the root as well.

File: src/libraries/System.Runtime.Extensions/tests/System/Random.cs [hunk 2]
```diff
@@ -606,43 +606,23 @@ public void Xoshiro_AlgorithmBehavesAsExpected()
...
+                Assert.Equal(0.02333050652594726, randOuter.NextDouble());
+                Assert.Equal(0.592402766665435, randOuter.NextDouble());
+                Assert.Equal(0.6658883755588669, randOuter.NextDouble());
+                Assert.Equal(0.18356709019700435, randOuter.NextDouble());
+                Assert.Equal(0.8686094193061149, randOuter.NextDouble());
+
                 Assert.Equal(0, randOuter.Next(0));
                 Assert.Equal(0, randOuter.Next(1));
 
-                Assert.Equal(11, randOuter.Next(42));
```
Comment (danmoseley) [inline src/libraries/System.Runtime.Extensions/tests/System/Random.cs:614]: any reason you removed all these cases, rather than simply updating to whatever the new algorithm produces?

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs [hunk 3]
```diff
@@ -90,46 +95,42 @@ public override int Next()
 
             public override int Next(int maxValue)
             {
-                if (maxValue > 1)
+                ulong randomProduct = (ulong)maxValue * NextUInt32();
+                uint lowPart = (uint)(randomProduct & uint.MaxValue);
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:99]: For those following along, this is implementing `m mod 2**32` with `m & (2**32-1)`

Which is more efficient than `m % 2**32` currently: https://github.com/dotnet/runtime/issues/11442#issuecomment-1363190619

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs [hunk 4]
```diff
@@ -90,46 +95,42 @@ public override int Next()
 
             public override int Next(int maxValue)
             {
-                if (maxValue > 1)
+                ulong randomProduct = (ulong)maxValue * NextUInt32();
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs:98]: I assume we are duplicating all this code from the other impl because the perf is significantly faster than calling through something. (This isn't new in this PR)

File: src/libraries/System.Runtime.Extensions/tests/System/Random.cs (hunk 2 above)
Comment (mla-alm) [reply src/libraries/System.Runtime.Extensions/tests/System/Random.cs:614]: Maybe I took the name of the test to literal: "Xoshiro_AlgorithmBehavesAsExpected"
From that I felt, that all it should do was to sufficiently verify, that the Xoshiro implementation was correct. From the referred link in the test https://www.pcg-random.org/posts/a-quick-look-at-xoshiro256.html it actually follows, that it is trivially predictable with only four outputs.

I could/can just put in the new values. But to me it is not a test of what the methods do, but of how they do it. It doesn't actually give any mark to whether or not the implementation is correct or not.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs (hunk 1 above)
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:34]: Done.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs (hunk 4 above)
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs:98]: I am not experienced at performance. But yes, I found a slight performance drop when having it call a common method. So I kind of assumed it was supposed to be duplicated.

**Comment by danmoseley** [comment]
Confirmed algorithms match the paper. Just a few comments.

BTW, looking at the Next() and NextInt64() methods, which you did not change, I wonder whether adjusting them as below produces slightly better assembly, what do you think?
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgAYBYAKDIEYbHqACV5mtzr4upZgGwgA7AObMAcjAQYAkkIyoAFAEoOXdQG816ncwDuACwCW/GM0UYoAVxiqWunVvsOHAelfMA4jAzMMEAAdmJABmZmAjDFw/CGYRH2ZsZgA3bH4bZiMhPwMzKGxRMwBtcjQBYREAOgBZbAQANTSbAF0y4CtfSwBPRJFsLO0XHXdMgDMcs1T0syNo7DAMKzT+HsFRGrrG6bK8CeZ8HwMIABNM6OOYUayYU/9mYgB2RJSmmEGhrhGsvfzC5hKymsqrUGq9lJV3h82FYgcxYLgYb4ALwSKQYACqcgUKBUzAAfHjmHQANyQqFjczwxHMACEKKBG1B0zs5M0ZNZrEe5iByip/AwpOcHOYAF92boxUKJYNxToeHxYZJpFjUMAVLLWE4OYYTGYLNZbBquFrhawRt5OoFgmEIlEYnEEkkphlvhhcnCCvF/qVyusQVsWm0On4oD1sH0BlLWV9xm7Jq8zokFkt+CtfcDNq8dtE4/tDidExcrkIbvauU7XkbdF9srnfl6AenGQHbBCo+SYRU4TAEfzmCilRiVTjlILTawjONFHzfHTmJ2/ZnmVXHCuXFzFDzpz3qQSiaO11xJRzj+TT6wmJwZe3ObxmJj5KhUdIH9jcRp7k9FAuRMpFAAlApjggfBKgAZQMbBYGOSpB2HXFR1Fa8gA===
```

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs (hunk 3 above)
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:99]: might be worth a comment.

File: src/libraries/System.Runtime.Extensions/tests/System/Random.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Runtime.Extensions/tests/System/Random.cs:614]: incidentally, these tests already force a certain seed for reproducibility, I wonder whether someone else who implemented this algorithm (eg Swift until the last PR) has a set of test values that we could copy to make certain we're doing the same thing.

Comment (danmoseley) [reply src/libraries/System.Runtime.Extensions/tests/System/Random.cs:614]: > But to me it is not a test of what the methods do, but of how they do it. It doesn't actually give any mark to whether or not the implementation is correct or not.

It covers boundary cases, eg 
```
               Assert.Equal(2147483646, randOuter.Next(int.MaxValue - 5, int.MaxValue));
                Assert.Equal(-1841045958, randOuter.Next(int.MinValue, int.MaxValue));
```
That's never a bad idea, when it can be done so trivially.

Comment (danmoseley) [reply src/libraries/System.Runtime.Extensions/tests/System/Random.cs:614]: It also seems we lose meaningful coverage of public API eg  `NextInt64()` and I don't think this is enough for the overload
```
                Assert.Equal(0, randOuter.NextInt64(0));
                Assert.Equal(0, randOuter.NextInt64(1));
```

Comment (danmoseley) [reply src/libraries/System.Runtime.Extensions/tests/System/Random.cs:614]: If it makes you feel better, you could think of them as protection against some future subtle codegen regression on some random architecture :)

**Comment by mla-alm** [comment]
> BTW, looking at the Next() and NextInt64() methods, which you did not change, I wonder whether adjusting them as below produces slightly better assembly, what do you think?
> 

From just looking at it, it does seem that it could be slightly better. Using dotnet/performance to benchmark the two gave me completely similar results.

**Comment by tkekalainen** [comment]
> BTW, looking at the Next() and NextInt64() methods, which you did not change, I wonder whether adjusting them as below produces slightly better assembly, what do you think? https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgAYBYAKDIEYbHqACV5mtzr4upZgGwgA7AObMAcjAQYAkkIyoAFAEoOXdQG816ncwDuACwCW/GM0UYoAVxiqWunVvsOHAelfMA4jAzMMEAAdmJABmZmAjDFw/CGYRH2ZsZgA3bH4bZiMhPwMzKGxRMwBtcjQBYREAOgBZbAQANTSbAF0y4CtfSwBPRJFsLO0XHXdMgDMcs1T0syNo7DAMKzT+HsFRGrrG6bK8CeZ8HwMIABNM6OOYUayYU/9mYgB2RJSmmEGhrhGsvfzC5hKymsqrUGq9lJV3h82FYgcxYLgYb4ALwSKQYACqcgUKBUzAAfHjmHQANyQqFjczwxHMACEKKBG1B0zs5M0ZNZrEe5iByip/AwpOcHOYAF92boxUKJYNxToeHxYZJpFjUMAVLLWE4OYYTGYLNZbBquFrhawRt5OoFgmEIlEYnEEkkphlvhhcnCCvF/qVyusQVsWm0On4oD1sH0BlLWV9xm7Jq8zokFkt+CtfcDNq8dtE4/tDidExcrkIbvauU7XkbdF9srnfl6AenGQHbBCo+SYRU4TAEfzmCilRiVTjlILTawjONFHzfHTmJ2/ZnmVXHCuXFzFDzpz3qQSiaO11xJRzj+TT6wmJwZe3ObxmJj5KhUdIH9jcRp7k9FAuRMpFAAlApjggfBKgAZQMbBYGOSpB2HXFR1Fa8gA===

Doesn't the second method return `long.MaxValue` when NextUInt64 returns `ulong.MaxValue - 1`?

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs [hunk 5]
```diff
@@ -90,46 +95,42 @@ public override int Next()
 
             public override int Next(int maxValue)
             {
-                if (maxValue > 1)
```
[the remark this answers is not in the document]
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs:93]: Firstly intention is for it to never receive a negative maxValue, right? As the ImplBase this is protected in System.Random:
image
Hence I assumed (perhaps wrongly) the check was merely there to avoid calculating the random in the trivial cases. I felt there was no need to make this check for all cases, just to save some calculation in the trivial cases.

What if it does receive a negative maxValue? Maybe (wrongly) used somewhere else internally... Since the algorithm actually continues with:
`uint usedMaxValue = (uint)maxValue`
in all places, and that it (in practice) will terminate for all uint values for this (imagined) usedMaxValue, I do not see a risk for a long nor an infinite loop for negative maxValue.

Currently it is not used anywhere else, as far as I can see.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs [hunk 6]
```diff
@@ -149,46 +150,42 @@ public override long NextInt64()
 
             public override long NextInt64(long maxValue)
             {
-                if (maxValue > 1)
+                UInt128 randomProduct = (UInt128)maxValue * NextUInt64();
```
[the remark this answers is not in the document]
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs]: Oversight = lack of knowledge. Thanks, added.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs [hunk 7]
```diff
@@ -149,46 +150,42 @@ public override long NextInt64()
...
+                ulong lowPart = (ulong)(randomProduct & ulong.MaxValue);
```
[the remark this answers is not in the document]
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs]: It compiled at ran fine locally without, so I assume it is not checked. Changed accordingly.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tkekalainen) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs:93]: You're right, my mistake. They shouldn't be negative. I forgot that calls to XoshiroImpl go through the methods implemented in Random because the methods in XoshiroImpl override ImplBase and not Random. Maybe a `Debug.Assert` that the inputs really are valid would be appropriate?

**Comment by danmoseley** [comment]
> Doesn't the second method return long.MaxValue when NextUInt64 returns ulong.MaxValue - 1?

You're right 😄

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs:93]: I think Debug.Asserts for non negative would be reasonable self-documentation.

[the remark this answers is not in the document]
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs:93]: Added.

File: src/libraries/System.Runtime.Extensions/tests/System/Random.cs (hunk 2 above)
Comment (mla-alm) [reply src/libraries/System.Runtime.Extensions/tests/System/Random.cs:614]: Reverted to old style.
Seems like the testing style for swift is a bit different: https://github.com/apple/swift/blob/e1d1760c878990c675aa9a30ce5bf145c2d8abbb/validation-test/stdlib/Random.swift

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs [hunk 8]
```diff
@@ -90,46 +95,46 @@ public override int Next()
...
                 }
 
-                Debug.Assert(maxValue == 0 || maxValue == 1);
-                return 0;
+                return (int)(randomProduct >> (sizeof(uint) * 8));
             }
 
             public override int Next(int minValue, int maxValue)
             {
+                Debug.Assert(minValue <= maxValue);
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:119]:
```suggestion
                Debug.Assert(minValue  >= 0 && minValue <= maxValue);
```
?

Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:119]: As far as I understand, minValue can be negative.
See also the tests/examples in Xoshiro_AlgorithmBehavesAsExpected (eg. https://github.com/dotnet/runtime/blob/f7633f498a8be34bee739b240a0aa9ae6a660cd9/src/libraries/System.Runtime.Extensions/tests/System/Random.cs#L622).

Comment (tkekalainen) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:119]: I don't see anything that says that negative values aren't allowed when both bounds are given. And the old implementations seem to work fine with negative bounds.

Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:119]: As long as we have tests then, great.

File: src/libraries/System.Runtime.Extensions/tests/System/Random.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Runtime.Extensions/tests/System/Random.cs:614]: One or two of those might be worth porting at some point, eg., the Chi squared test, not suggesting you do it.

Historical context, the original/existing System.Random algorithm (still used for seeded constructor) as well as being very slow is also flawed due to an error made in the original code. This has been reported many times, with long discussion [here for example](https://github.com/dotnet/runtime/issues/23198). The decision was made to keep it for compatibility, as it's OK-ish, and fortunately the unseeded constructor is the most commonly used, so we were able to replace with a much faster algorithm producing a better distribution. I don't have easy access to whatever tests were written for System.Random (this was 20 years ago), but my guess is that they were simple functional tests, and not statistical tests, since the algorithm was just copied from Knuth (and assumed accurately so). I wonder whether some basic statistical tests could have discovered the bug way back then.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs (hunk 8 above)
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:119]: I see some basic tests for it.

File: src/libraries/System.Runtime.Extensions/tests/System/Random.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Runtime.Extensions/tests/System/Random.cs:614]: I see Golang tests do statistical checks as well
https://github.com/golang/go/blob/master/src/math/rand/rand_test.go

**Comment by danmoseley** [comment]
I have no more feedback, but we should have someone from @dotnet/area-system-runtime or @stephentoub review before merging.

We'll likely need perf results (relative to current) on the reviewed sources, before merging.

thanks for working on this! will be a nice addition to the .NET 8 performance blog post a year from now 😄

**Comment by danmoseley** [comment]
I experimented with a chi-squared test. The Swift approach of bucketing into 50 buckets works pretty well. It's effective at testing for off by ones if you pass it a relatively small range. 

Of course if you're testing that you reach the whole range, if you generate enough values on a narrow range, you can confidently enough assert that you see at least one sample from the ends of the range. The Chi-squared does have the advantage of also flagging certain quite bad distributions, but that's possibly not terribly interesting since we're not trying to test the generator itself.

https://gist.github.com/danmoseley/784a25839978425ae15d037f587ac76b

Maybe I'll add this as a separate PR later.

**Comment by jeffhandley** [comment]
@mla-alm I wanted to confirm that the existing dotnet/performance benchmarks were able to observe the effects of this change and that no new benchmarks are needed.

Can you share the results of baseline vs. pr from your local runs of the applicable benchmarks as comment here please?

Thanks!

**Review by danmoseley — APPROVED** [review]
Signing off based on reviewing it last year, but as mentioned someone in the libraries team should sign off before merging as it's "fiddly"

I would also be interested in their opinion on whether I should throw up a PR for these statistical tests (gist above)

**Comment by mla-alm** [comment]
@jeffhandley Yes, the current benchmarks gave results like the following when I ran it:

No Slower results for the provided threshold = 2% and noise filter = 0.3 ns.
| Faster                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| ------------------------------------------------ | ---------:| ----------------:| ----------------:| --------:|
| System.Tests.Perf_Random.Next_int_int_unseeded   |      3.25 |             6.66 |             2.05 |         |
| System.Tests.Perf_Random.Next_int_unseeded       |      3.23 |             6.38 |             1.97 |         |
| System.Tests.Perf_Random.Next_long_long_unseeded |      2.66 |             6.48 |             2.43 |         |
| System.Tests.Perf_Random.Next_long_unseeded      |      1.45 |             3.84 |             2.64 |         |


Whether new benchmarks are needed depends on a lot of factors. I would personally go with that they cover it "just fine" for now.

**Comment by jeffhandley** [comment]
> I would also be interested in their opinion on whether I should throw up a PR for these statistical tests (gist above)

@danmoseley Yes, that would be great to have as a separate PR. Thanks for the offer to do that.

**Comment by danmoseley** [comment]
OK will do after this goes in so as not to conflict.

**Comment by jeffhandley** [comment]
The `Build Browser wasm Linux Release LibraryTests_AOT` leg that failed is catching my attention here, but the build logs have been purged and I cannot get the leg to rerun. I'm merging main into your fork/branch, @mla-alm. That should help us get a fresh run from CI.

**Review by jeffhandley — APPROVED** [review]
This looks good to me too. There's one check that is timing out, but everything else is green.

Let's wait for @stephentoub and/or @tannergooding to be able to review before merging though. It might be next week or the week after before one of them can review it.

Thank you, @mla-alm!

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs [hunk 9]
```diff
@@ -149,56 +154,46 @@ public override long NextInt64()
...
-                            return (long)result + minValue;
-                        }
+                        randomProduct = (UInt128)exclusiveRange * NextUInt64();
+                        lowPart = (ulong)randomProduct;
                     }
                 }
 
-                Debug.Assert(minValue == maxValue || minValue + 1 == maxValue);
-                return minValue;
+                return (long)(randomProduct >> (sizeof(ulong) * 8)) + minValue;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:196]: Same here, I think `64` would be clearer than `sizeof(ulong) * 8)`

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs [hunk 10]
```diff
@@ -90,46 +95,46 @@ public override int Next()
...
+                ulong randomProduct = (ulong)(uint)maxValue * NextUInt32();
+                uint lowPart = (uint)randomProduct;
+
+                if (lowPart < (uint)maxValue)
                 {
-                    // Narrow down to the smallest range [0, 2^bits] that contains maxValue.
-                    // Then repeatedly generate a value in that outer range until we get one within the inner range.
-                    int bits = BitOperations.Log2Ceiling((uint)maxValue);
-                    while (true)
+                    uint remainder = ((uint)0 - (uint)maxValue) % (uint)maxValue;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs:105]: `(uint)0` can be `0u`, `(ulong)0` can be `0ul`, etc.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs [hunk 11]
```diff
@@ -90,46 +95,46 @@ public override int Next()
 
             public override int Next(int maxValue)
             {
-                if (maxValue > 1)
+                Debug.Assert(maxValue >= 0);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:98]: It looks like this implementation could just be `public override int Next(int maxValue) => Next(0, maxValue);`, and effectively the only additional cost would be one subtraction and one addition.  Can we avoid the duplication by doing so?  Is the difference measurable?

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs [hunk 12]
```diff
@@ -90,46 +95,46 @@ public override int Next()
...
-                            return (int)result + minValue;
-                        }
+                        randomProduct = (ulong)exclusiveRange * NextUInt32();
+                        lowPart = (uint)randomProduct;
                     }
                 }
 
-                Debug.Assert(minValue == maxValue || minValue + 1 == maxValue);
-                return minValue;
+                return (int)(randomProduct >> (sizeof(uint) * 8)) + minValue;
```
[the remark this answers is not in the document]
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs]: Fixed.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs (hunk 9 above)
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs]: Fixed.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs (hunk 10 above)
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro256StarStarImpl.cs]: Fixed.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs (hunk 11 above)
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:98]: Yes, this could be done. As I recall it added around 0.3 ns to the benchmarks for me. Seemed important at the time. Not so much anymore. Should it be changed?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:98]: Actually, we should be able to write just two static utility methods, one that takes a uint max and one that takes a ulong max, and then use those two routines from the four methods being modified here across both version of xoshiro. There shouldn't be any additional cost if those two helpers are aggressive inlining.

Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:93]: The methods should also be able to access the generator methods - otherwise the cost increased. Tried passing the generator methods as arguments to static methods, but this appeared to be even more expensive.

I have now added a common abstract base class with the implementation logic. On my system this gives approximately the same results as before. Will this be an acceptable solution?

I guess one could discuss the layers of the classes. I went with minimal intervention to make it work, and hence putting LemireImpl as base to XoshiroImpl.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs [hunk 13]
```diff
@@ -90,46 +95,46 @@ public override int Next()
 
             public override int Next(int maxValue)
             {
-                if (maxValue > 1)
+                Debug.Assert(maxValue >= 0);
+
+                ulong randomProduct = (ulong)(uint)maxValue * NextUInt32();
```
[the remark this answers is not in the document]
Comment (mla-alm) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs]: True, I have only tested on 64-bit.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs (hunk 11 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:93]: Thanks.  I was thinking more like this: https://github.com/dotnet/runtime/compare/main...stephentoub:runtime:randomrange

**Review by stephentoub — COMMENTED** [review]
The implementation and 64-bit perf look good. I'm building locally for 32-bit to check that.

**Comment by stephentoub** [comment]
>> This Xoshiro128 implementation is only used on 32-bit. I saw some measurements were done, but I believe those were on 64-bit... have we measured how this performs on 32-bit?

> True, I have only tested on 64-bit.

On 32-bit, I get numbers like this:

|                  Method |             Toolchain |      Mean |     Error |    StdDev |    Median |       Min |       Max | Ratio |
|------------------------ |---------------------- |----------:|----------:|----------:|----------:|----------:|----------:|------:|
|       Next_int_unseeded | \main_x86\corerun.exe | 13.130 ns | 0.4286 ns | 0.4936 ns | 12.949 ns | 12.475 ns | 14.104 ns |  1.00 |
|       Next_int_unseeded |   \pr_x86\corerun.exe |  5.612 ns | 0.0956 ns | 0.0847 ns |  5.602 ns |  5.489 ns |  5.813 ns |  0.43 |
|                         |                       |           |           |           |           |           |           |       |
|   Next_int_int_unseeded | \main_x86\corerun.exe | 12.036 ns | 0.1935 ns | 0.1616 ns | 11.969 ns | 11.864 ns | 12.342 ns |  1.00 |
|   Next_int_int_unseeded |   \pr_x86\corerun.exe |  5.296 ns | 0.0605 ns | 0.0565 ns |  5.288 ns |  5.208 ns |  5.423 ns |  0.44 |
|                         |                       |           |           |           |           |           |           |       |
|      Next_long_unseeded | \main_x86\corerun.exe | 11.874 ns | 0.2669 ns | 0.2496 ns | 11.850 ns | 11.462 ns | 12.305 ns |  1.00 |
|      Next_long_unseeded |   \pr_x86\corerun.exe | 14.409 ns | 0.1976 ns | 0.1650 ns | 14.360 ns | 14.121 ns | 14.758 ns |  1.22 |
|                         |                       |           |           |           |           |           |           |       |
| Next_long_long_unseeded | \main_x86\corerun.exe | 15.087 ns | 0.2617 ns | 0.2320 ns | 15.085 ns | 14.771 ns | 15.666 ns |  1.00 |
| Next_long_long_unseeded |   \pr_x86\corerun.exe | 18.001 ns | 0.2120 ns | 0.1983 ns | 17.967 ns | 17.711 ns | 18.399 ns |  1.19 |

So, at least according to these benchmarks, this PR represents a potential regression on 32-bit.

Can you measure on 32-bit as well, @mla-alm?

**Comment by mla-alm** [comment]
I am running on Linux, so cannot measure on 32-bit.

`public long Next_long_long_unseeded() => _randomUnseeded.NextInt64(100, 10000);`
So this benchmark for the 32-bit is in reality only handling the case of an int on the 'main' test.

Maybe it is better to be conservative and leave the implementation (at least the NextInt64) as is in Xoshiro128?

**Comment by jeffhandley** [comment]
> Maybe it is better to be conservative and leave the implementation (at least the NextInt64) as is in Xoshiro128?

@stephentoub What do you think about that idea?

**Comment by stephentoub** [comment]
Yeah, let's leave the Xoshiro128's NextInt64 unmodified for now, other than a comment indicating why it's not getting the same treatment as the others.

**Comment by mla-alm** [comment]
Xoshiro128's NextInt64 is as it was before. And comment added.

File: src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs [hunk 14]
```diff
@@ -147,6 +117,9 @@ public override long NextInt64()
                 }
             }
 
+            // NextInt64 in Xoshiro128 has not been implemented with the Lemire algorithm like the related methods.
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs:120]:
```suggestion
            // NextInt64 in Xoshiro128 has not been implemented with the fastrange algorithm like the related methods.
```

**Review by stephentoub — APPROVED** [review]
Thanks!

**Comment by danmoseley** [comment]
@mla-alm Thank you! Perhaps you'd be interested in another contribution somewhere?

**Comment by EgorBo** [comment]
Improvements on x64 https://github.com/dotnet/perf-autofiling-issues/issues/13151

**Comment by mla-alm** [comment]
Hi @danmoseley.
Sure I might. Do you have any suggestions?
Otherwise I might look around, when I find the time.

**Comment by EgorBo** [comment]
Improvements on win-x64:
* https://github.com/dotnet/perf-autofiling-issues/issues/13255
